### PR TITLE
Anonymous user upgrade federated

### PR DESCRIPTION
**Pull request by protocol86**
Anonymous user upgrade federated accounts and phone auth. 

Swift sample will be updated in a  follow up PR.

Addresses #139

### CODE REVIEW

File: FirebaseAuthUI/FUIAuth.h [hunk 1]
```diff
@@ -190,6 +190,11 @@ __attribute__((deprecated("Instead use authUI:didSignInWithAuthDataResult:error:
  */
 @property(nonatomic, copy, nullable) NSURL *TOSURL;
 
+/** @property autoUpgradeAnonymousUsers
+    @brief Whether to enable auto upgrading of anonymous accounts, defaults to NO.
+ */
+@property(nonatomic, assign, getter=isAutoUpgradeAnonymousUsers) BOOL autoUpgradeAnonymousUsers;
```
Comment (morganchen12) [inline FirebaseAuthUI/FUIAuth.h]: nit: rename this to `shouldAutoUpgradeAnonymousUsers` (for both getter and setter)

File: samples/objc/FirebaseUI-demo-objc/Samples/Auth/FUIAuthViewController.m [hunk 2]
```diff
@@ -279,21 +310,9 @@ - (NSString *)getAllIdTokens {
 
 - (void)signOut {
   NSError *error;
-  FIRUser *currentUser = self.authUI.auth.currentUser;
-  // If the user is anonymous, delete the user to avoid dangling anonymous users.
-  if (currentUser.isAnonymous) {
-    [currentUser deleteWithCompletion:^(NSError * _Nullable error) {
```
Comment (morganchen12) [inline samples/objc/FirebaseUI-demo-objc/Samples/Auth/FUIAuthViewController.m]: Do we no longer need to delete anonymous users?

Comment (protocol86) [reply samples/objc/FirebaseUI-demo-objc/Samples/Auth/FUIAuthViewController.m]: We do, I just moved the code to the tableview cell. The reason is that I wanted to use the sign in button to trigger a sign-in when the current user was anonymous. Essentially I am separating anonymous sign in/out with "normal" sign in/out.

**Comment by morganchen12** [comment]
#403 is now merged, feel free to merge this PR as soon as travis tests pass.

Thanks!

File: FirebaseAuthUI/FUIAuth.h (hunk 1 above)
Comment (protocol86) [reply FirebaseAuthUI/FUIAuth.h]: Thanks, done.

**Pull request merged**